File: transformation/src/many/src/many_s.cpp
#include "ros/ros.h"
#include "tf2_ros/transform_listener.h"//创建订阅对象所需包含的头文件
#include "tf2_ros/buffer.h"
#include "geometry_msgs/PointStamped.h"//创建坐标点所需包含的头文件
#include "tf2_geometry_msgs/tf2_geometry_msgs.h"
#include "geometry_msgs/TransformStamped.h"//运用坐标系之间关系所需包含的头文件

/*

需求:
    现有坐标系统，父级坐标系统 world,下有两子级系统 son1，son2，
    son1 相对于 world，以及 son2 相对于 world 的关系是已知的，
    求 son1 与 son2中的坐标关系，又已知在 son1中一点的坐标，要求求出该点在 son2 中的坐标
实现流程:
    1.包含头文件
    2.初始化 ros 节点
    3.创建 ros 句柄
    4.创建 TF 订阅对象
    5.解析订阅信息中获取 son1 坐标系原点在 son2 中的坐标
      解析 son1 中的点相对于 son2 的坐标
    6.spin
*/
int main(int argc,char*argv[])
{
    setlocale(LC_ALL,"");
    ros::init(argc,argv,"many_s");//初始化 ros 节点
    ros::NodeHandle nh();//创建 ros 句柄
    //创建订阅对象
    tf2_ros::Buffer buffer;
    tf2_ros::TransformListener sub(buffer);
    //创建坐标点
    geometry_msgs::PointStamped psSon1;
    psSon1.header.stamp=ros::Time::now();//编写该坐标点具体信息
    psSon1.header.frame_id="son1";
    psSon1.point.x=1.0;
    psSon1.point.y=2.0;
    psSon1.point.z=3.0;
    //编写解析逻辑
    ros::Rate rate(10);//以1秒10次进行接收发布方的数据
    while(ros::ok())
    {
        try
        {
            //计算son1与son2的相对关系
            geometry_msgs::TransformStamped son1to2=buffer.lookupTransform("son2","son1",ros::Time(0));//目标坐标系，原坐标系，取两个坐标系最近一次的时间计算相对关系
            ROS_INFO("son1相对于son2的信息:父坐标系名称:%s,子坐标系名称:%s,偏移量信息:(%.2f,%.2f,%.2f)",son1to2.header.frame_id.c_str(),son1to2.child_frame_id.c_str(),son1to2.transform.translation.x,son1to2.transform.translation.y,son1to2.transform.translation.z);
            //计算son1中的某个坐标点在son2的坐标值
            geometry_msgs::PointStamped psSon2=buffer.transform(psSon1,"son2");//被转换的坐标点，目标坐标系
            ROS_INFO("坐标点在son2中的值:(%.2f,%.2f,%.2f)",psSon2.point.x,psSon2.point.y,psSon2.point.z);
        }
        catch(const std::exception &e)
        {
            ROS_INFO("错误信息为:%s",e.what());
        }

        rate.sleep();
        ros::spinOnce();
    }
}

File: transformation/src/many/scripts/many_s.py
#! /usr/bin/env python
import rospy
import tf2_ros#创建订阅对象所需包含的头文件
from tf2_geometry_msgs import tf2_geometry_msgs#组织被转换的坐标点所需包含的头文件
from geometry_msgs import TransformStamped
"""  
    订阅坐标系信息，生成一个相对于 子级坐标系的坐标点数据，
    转换成父级坐标系中的坐标点

    实现流程:
        1.导包
        2.初始化 ROS 节点
        3.创建 TF 订阅对象
        4.创建一个 radar 坐标系中的坐标点
        5.调研订阅对象的 API 将 4 中的点坐标转换成相对于 world 的坐标
        6.spin

"""


if __name__=="__main__":
    rospy.init_node("static_sp")#初始化节点
    #创建订阅对象
    buffer=tf2_ros.Buffer()#创建缓存对象
    sub=tf2_ros.TransformListener(buffer)#创建订阅对象（将所订阅的数据传入缓存中）
    #组织被转换的坐标点
    ps=tf2_geometry_msgs.PointStamped()#组织被发布消息【参看坐标msgs消息】
    ps.header.stamp=rospy.Time.now()
    ps.header.frame_id="son1"#参考坐标点
    ps.point.x=1.0
    ps.point.y=2.0
    ps.point.z=3.0
    #转换逻辑实现，调用tf封装的算法
    # rospy.sleep(2)#通过休眠以接受完发布方的数据
    rate=rospy.Rate(10)#以1秒10次去接受发布方的数据
    while not rospy.is_shutdown():
        try:
            #计算son1相对于son2的坐标关系
            ts=buffer.lookup_transform("son2","son1",rospy.Time(0))
            rospy.lohginfo("父系坐标系名称:%s ,子系坐标系名称:%s ,偏移值:(%.2f,%.2f,%.2f)",ts.header.frame_id,ts.child_frame_id,ts.transform.translation.x,ts.transform.translation.y,ts.transform.translation.z)
            #转换实现
            ps_out=buffer.transform(ps,"son2")#被转换的坐标点、目标坐标系
            #输出结果
            rospy.loginfo("转换后的坐标值:(%.2f,%.2f,%.2f),参考坐标系:%s",ps_out.point.x,ps_out.point.y,ps_out.point.z,ps_out.header.frame_id)
        except Exception as e:
            rospy.logwarn("错误信息:%s",e)
        
        rate.sleep()


    
    
